app/models/address.rb

```diff
@@ -0,0 +1,31 @@
+class Address < ApplicationRecord
+  # The REGEXP for the postcode validation is taken from this very detailed stack overflow answer:
+  # https://stackoverflow.com/questions/164979/uk-postcode-regex-comprehensive/51885364#51885364
+  # The second REGEXP from the answer section of that comment was used because as part
+  # of the validation process we are changing postcode to uppercase and removing spaces this allows
+  # for some simplification of the regular expression
+
+  POSTCODE_REGEXP = /([A-Z][A-HJ-Y]?[0-9][A-Z0-9]? ?[0-9][A-Z]{2}|GIR ?0A{2})/
```

Comment:
The regexp should enforce that the postcode **starts** (`\A`) and **ends** (`\z`) with the defined pattern. As it stands, something like `bla bla SW1H 9AJ bla bla`:

<img width="1062" alt="screen shot 2018-10-17 at 09 24 35" src="https://user-images.githubusercontent.com/67125/47072882-ef2e2300-d1ee-11e8-8aa3-cdf3bb2e42eb.png">

whereas the restricted version does match it correctly:

<img width="1071" alt="screen shot 2018-10-17 at 09 26 59" src="https://user-images.githubusercontent.com/67125/47072960-184eb380-d1ef-11e8-8080-6ff718b6e41b.png">

<img width="1028" alt="screen shot 2018-10-17 at 09 26 39" src="https://user-images.githubusercontent.com/67125/47072977-20a6ee80-d1ef-11e8-8dc3-ac6c426ca7ee.png">

Reply:
You can find more info on this [here](https://ruby-doc.org/core-2.5.1/Regexp.html#class-Regexp-label-Anchors).